### ISSUE TITLE
Faster packing / unpacking of variable length sequences

### ISSUE THREAD

**Issue by elanmart** [pytorch/pytorch#1788]
@jekbradbury mentioned on the Forums that it would be possible to speed up packing and unpacking variable length sequences, by moving the code to C++

In my current work I find this packing really helpful, but it is also a performance bottleneck.

I'd like to ask few questions regarding this:

1. Do you know if anyone is already working on this feature?
2. If not, could anyone please advice me how I could best approach this problem, given I have some prior experience working with python C-API?
3. If I were allowed to submit it, would You be interested in a PR implementing this feature?

**Comment by apaszke** [comment]
Is it a performance bottleneck because of the copies or because of Python? Did you check?

**Comment by apaszke** [comment]
We'll be happy to provide help and accept patches to make it faster

**Comment by stefbraun** [comment]
@apaszke Here is a script that provides two ways of using packed sequences: one way is fast, the other way is around 4 times slower. Both yield the same checksums at the final layer output. I am not able to explain why one is faster, could you please have a look at it?

Output of fast variant
```
fast_packed (
  (gru): GRU(123, 320)
  (linear): Linear (320 -> 10)
)
::: # network parameters: 430410
::: Checksum of final layer output from epoch 0:8.07597064972
::: Checksum of final layer output from epoch 5:6.69699001312
::: Checksum of final layer output from epoch 10:1.61638641357
::: Checksum of final layer output from epoch 15:-6.10106754303
::: Checksum of final layer output from epoch 20:-11.1396369934
>>> Median runtime per epoch [sec] 0.222762107849
```
Output of slow variant
```
slow_packed (
  (gru): GRU(123, 320)
  (linear): Linear (320 -> 10)
)
::: # network parameters: 430410
::: Checksum of final layer output from epoch 0:8.07597064972
::: Checksum of final layer output from epoch 5:6.69699001312
::: Checksum of final layer output from epoch 10:1.61638641357
::: Checksum of final layer output from epoch 15:-6.10106754303
::: Checksum of final layer output from epoch 20:-11.1396369934
>>> Median runtime per epoch [sec] 0.930104017258
```
Script
```
from timeit import default_timer as timer

import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.autograd import Variable
from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence, PackedSequence

# Set seeds
np.random.seed(11)
torch.manual_seed(11)
torch.cuda.manual_seed(11)


# Provide default parameters
def default_params():
    rnn_size = 320
    learning_rate = 1e-3
    epochs = 25
    return rnn_size, learning_rate, epochs


# Provide a toy batch
def toy_batch(seed=11, shape=(25, 1000, 123), classes=10):
    batch_size, max_len, features = shape
    np.random.seed(seed)

    # Samples
    bX = np.float32(np.random.uniform(-1, 1, (shape)))
    b_lenX = np.int32(np.linspace(max_len, max_len, batch_size))
    # print('::: Lengths of samples in batch: {}'.format(b_lenX))

    # Targets
    bY = np.int32(np.random.randint(low=0, high=classes - 1, size=batch_size))

    return bX, b_lenX, bY, classes


# Get data
bX, b_lenX, bY, classes = toy_batch()
batch_size, max_len, features = bX.shape
rnn_size, learning_rate, epochs = default_params()

# PyTorch compatibility: time first, batch second
bX = np.transpose(bX, (1, 0, 2))

# Create symbolic vars
bX = Variable(torch.from_numpy(bX).cuda())
bX = pack_padded_sequence(bX, b_lenX[::-1])  # Pack those sequences for masking, plz

bY = Variable(torch.from_numpy(bY).cuda())


# Network 1: slow use of packed sequences
class slow_packed(nn.Module):
    def __init__(self, features, rnn_size):
        super(slow_packed, self).__init__()
        self.gru = nn.GRU(input_size=features, hidden_size=rnn_size, bias=True)
        self.linear = nn.Linear(rnn_size, classes, bias=True)

    def forward(self, x):
        h1p, state = self.gru(x)  # RNN
        h1, lens = pad_packed_sequence(h1p)  # unpack
        max_len, batch_size, features = h1.size()  # get sizes for reshape
        h2 = h1.view(max_len * batch_size, -1)  # reshape
        h3 = self.linear(h2)  # linear transform
        h4 = h3.view(max_len, batch_size, -1)  # reshape
        h5 = h4[-1, :, :]  # slice last element
        return h5


# Network 2: fast use of packed sequences
class fast_packed(nn.Module):
    def __init__(self, features, rnn_size):
        super(fast_packed, self).__init__()
        self.gru = nn.GRU(input_size=features, hidden_size=rnn_size, bias=True)
        self.linear = nn.Linear(rnn_size, classes, bias=True)

    def forward(self, x):
        h1p, state = self.gru(x)  # RNN
        h3 = self.linear(h1p.data)  # linear transform
        h3p = PackedSequence(h3, h1p.batch_sizes)  # create PackedSequence (pytorch docs: Instances of this class should never be created manually. They are meant to be instantiated by functions like pack_padded_sequence().
        h4, lens = pad_packed_sequence(h3p)  # unpack
        h5 = h4[-1, :, :]  # slice last element
        return h5


########################################################
# Switch between fast and slow variant by commenting
net = slow_packed(features=features, rnn_size=rnn_size)
# net = fast_packed(features=features, rnn_size=rnn_size)
########################################################
net.cuda()  # move network to GPU
print(net)

# Print parameter count
params = 0
for param in list(net.parameters()):
    sizes = 1
    for el in param.size():
        sizes = sizes * el
    params += sizes
print('::: # network parameters: ' + str(params))

# Create optimizer, criterion
optimizer = optim.Adam(net.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()  # loss definition

# Start training
time = []
for ep in range(epochs):

    start = timer()
    optimizer.zero_grad()
    output = net(bX)
    loss = criterion(output, bY.long())
    loss.backward()
    optimizer.step()

    torch.cuda.synchronize()  # Synchronize for more precise timing
    end = timer()
    time.append(end - start)

    if ep % 5 == 0:
        print('::: Checksum of final layer output from epoch {}:{}'.format(ep, torch.sum(output).cpu().data.numpy()[0]))

print('>>> Median runtime per epoch [sec] {}'.format(np.median(time)))
```

**Comment by apaszke** [comment]
The two examples do completely different things. The slow network transposes the full RNN output (320 in feature dim), while the fast one transposes the output of the Linear layer (10 in feature dim). It's 32x less data to shuffle, so run times look ok. I think it's reasonable to guarantee that you can apply pointwise ops and e.g. Linear to packed sequences, but it seems orthogonal to this issue, which proposes to improve the implementation of packed sequence function. If you have any proposals, please open a new issue, or send PRs!
 
Also, I think your slicing might be incorrect. Taking -1 from each sample might slice off padding.

**Comment by stefbraun** [comment]
thanks for your answer, to clarify:
- every sample has the same sequence length here, in order to be able to compare the checksums of the final layers output. This makes slicing by -1 ( at max length = 1000) feasible.
- Both networks should do the same thing, as the checksum of the final layer is the same. It is common to put a dense layer on top of the RNN in speech recognition. This dense layer maps the output of each time step to classes in the case of CTC.
- As the dense layer operates on each time step, it needs an input of (max_len*batch_size, rnn_size). I can either do that by viewing the unpacked RNN output (```h2``` from the slow example) OR I can directly use the data from the packed sequence (```h1p.data``` from the fast example).

**Comment by apaszke** [comment]
They are equivalent semantically, so they give the same outputs. But by changing the order of operations you can dramatically change the complexity of it. As an example consider this matrix multiplication sequence: [1024 x 2] * [2 x 1024] * [1024 * 1024]. Now, you can order the operations in two ways:
* ([1024 x 2] * [2 x 1024]) * [1024 * 1024] -> [1024 x 1024] * [1024 x 1024]
* [1024 x 2] * ([2 x 1024] * [1024 * 1024]) -> [1024 x 2] * [2 x 1024]
Both orderings will give you the same outputs, but second one will be waaaay faster to compute. This is exactly what happens here. They both compute the same thing, just swap the order of some associative operations.

**Comment by stefbraun** [comment]
Actually, are you saying that unpacking (25000,10) is so much faster than unpacking (25000,320) that it plays a role in this case?

From my understanding, the GRU and dense layers do the same computation on the same tensors. Consequently, the speed difference should come from the packing functions. I put it into a table to highlight the flow. Perhaps this could be a recommended workflow until packing/unpacking speed is improved.

![image](https://user-images.githubusercontent.com/13469638/27330956-bf1bd154-55bb-11e7-8f7c-ad4e48667a69.png)

**Comment by apaszke** [comment]
Yes. As I said, it's a 32x difference in the amount of bytes that you need to rearrange. GRU and Dense layers do the same jobs.

**Comment by elanmart** [comment]
Hi @apaszke , apologies for being terrible at communication!

We've decided to go with a different model after all, and I did not have time to further investigate the issue. 

But I wanted to follow-up on this on my own. 

Initially I thought the problem is due to the packing / unpacking being done in Python,
but from some tests I did today it seems like it's not the case. 

For an epoch of training, packing & unpacking takes `~3s`, and running LSTM `~10s`. But it seems like the biggest penalty is due to autograd -- with packing, calling `backward` takes `~50s`, while without it's around `~18`. 

I've tested this with sequences of equal length, and for ones sampled randomly from `(1, 128)`, and the results are similar.

I'll try to follow-up with more details once I get the time to benchmark this with a better tool than `time.perf_counter()`. I'll post the script and more results when I double-check the script.

**Comment by soumith** [comment]
@prigoyal wanna take this?
We need to move `pack_padded_sequence` and `pad_packed_sequence` to the C++ side of things.

**Comment by dirkweissenborn** [comment]
Hi, I struggled with the same issue and found that the python code for packing and unpacking is highly unoptimized since it slices tensors for each timestep, which is really unecessary. Consider the case where a lengths in the batch are the same, a single slice and reshaping would be enough. Basically we only need to iterate over the different batch lengths we have and NOT over all timesteps which is really wasteful. Similarly, I think padding could be improved.

 I quick fix for improving packing would be the following:

```python
def pack_padded_sequence(input, lengths, batch_first=False):
    if lengths[-1] <= 0:
        raise ValueError("length of all samples has to be greater than 0, "
                         "but found an element in 'lengths' that is <=0")
    if batch_first:
        input = input.transpose(0, 1)

    steps = []
    batch_sizes = []
    lengths_iter = reversed(lengths)
    batch_size = input.size(1)
    if len(lengths) != batch_size:
        raise ValueError("lengths array has incorrect size")

    prev_l = 0
    for i, l in enumerate(lengths_iter):
        if l > prev_l:
            c_batch_size = batch_size - i
            steps.append(input[prev_l:l, :c_batch_size].contiguous().view(-1, input.size(2)))
            batch_sizes.extend([c_batch_size] * (l - prev_l))
            prev_l = l

    return nn.utils.rnn.PackedSequence(torch.cat(steps), batch_sizes)
```

**Comment by apaszke** [comment]
@dirkweissenborn makes sense, can you please send a PR?

**Comment by dirkweissenborn** [comment]
@apaszke done

**Comment by soumith** [comment]
partly fixed via #3278 
We can improve the runtime further by moving the call into C

**Comment by stefbraun** [comment]
As described in the above [comment](https://github.com/pytorch/pytorch/issues/1788#issuecomment-325826495)  the main problem lies in using packing/unpacking during the backwards pass. Does [#3278](https://github.com/pytorch/pytorch/pull/3278) actually help during the backwards pass?

**Comment by apaszke** [comment]
@stefbraun it should help a bit, but there are still things that could be improved

**Comment by dirkweissenborn** [comment]
unpacking is not affected, but packing should be much faster now in the backward pass.

**Comment by apaszke** [comment]
It's not the matter of unpacking. Your PR reduced the number of narrows we have, which will make backwards faster. Actually I just realized that just putting the code in an autograd function, and implementing its backward manually will probably make it a few times faster

**Comment by dirkweissenborn** [comment]
@apaszke yes, there are far less torch ops now. I did something similar for padding. However, having a Function for it might be best.

**Comment by apaszke** [comment]
Having a function will help a lot. I just realized what makes the backward so slow - every `narrow` actually creates a tensor with size equal to the one that was narrowed in the forward, fills it with 0s, and copies the grad into the narrowed slice. Since we're taking a lot of slices, and the gradient is very sparse, a lot of time is spent on this zeroing

**Comment by ozancaglayan** [comment]
Hello,

I also found it very unintuitive that the method requires a Python list for `lengths` and fails if given a `TorchTensor`.  Would it be bad to add a `.tolist()` call if the given object is a TorchTensor?

**Comment by apaszke** [comment]
@ozancaglayan I think a better way to handle it would be to rewrite it in a way such that it works with both lists and tensors. It's mostly indexing/iteration so it should be easy

**Comment by elanmart** [comment]
Wouldn't the backward for `pack_padded_sequence` be as simple the following? :
```python
@staticmethod
def backward(ctx, grad_data, grad_batch_sizes):

    batch_sizes, = ctx.saved_variables
    input_size   = ctx.input_size
    
    ret = grad_steps.new(*input_size).zero_()
    idx = 0

    for i, val in enumerate(batch_sizes.data):
        val = int(val)

        ret[i, :val, ...] = grad_data[idx:idx+item, ...]
        idx += val

    return ret, None
```

I've needed the `packing` today again, and simply providing this `backward` reduces `CPU` backward from `600ms` to `14ms`, and it seems to return the same gradients as calling current implementation.

Could I open a PR with this? Also, should this code go to `python` sources or to `ATen`? I've implemented it in `ATen` also but didn't find it to give any speedups (for the simple benchmarks I run)

**Comment by elanmart** [comment]
@soumith perhaps this can be closed thanks to https://github.com/pytorch/pytorch/pull/4512 ?
